libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -713,20 +712,37 @@ float AC_AttitudeControl::get_althold_lean_angle_max() const
 }
 
 // Proportional controller with piecewise sqrt sections to constrain second derivative
-float AC_AttitudeControl::sqrt_controller(float error, float p, float second_ord_lim)
+float AC_AttitudeControl::sqrt_controller(float error, float p, float second_ord_lim, float dt)
 {
-    if (second_ord_lim < 0.0f || is_zero(second_ord_lim) || is_zero(p)) {
-        return error*p;
+    float correction_rate;
+    if (second_ord_lim < 0.0f || is_zero(second_ord_lim)) {
+        // second order limit is zero or negative.
+        correction_rate = error*p;
+    } else if (is_zero(p)) {
+        // P term is zero but we have a second order limit.
+        if (error > 0.0f) {
+            correction_rate = safe_sqrt(2.0f*second_ord_lim*(error));
+        } else if (error < 0.0f) {
```

Comment:
In all three above, I think it would be better to use:

- `!is_positive(second_ord_lim)`
- `is_positive(error) && ...`
- `is_negative(error) && ...`

Reply:
great, thanks for the review, I'll fix this.

Reply:
I'm replacing "error > 0.0f" with "is_positive(error)", etc so that's slightly different that you suggested but I think it's correct.